### PR TITLE
OJ-1922-Add-validator-for-NINO

### PR DESCRIPTION
**Pull request by liamStockwellGDS**
adding validation to the Nino page.
OJ-1922

### What changed

added the validator, and attached it to the field
[update to field and validator
nino validator edge case unit tests
front-end testing for the nino validation.

### Why did it change

so that the UI has validation

### Issue tracking

[OJ-1922] - https://govukverify.atlassian.net/browse/OJ-1922

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

[OJ-1922]: https://govukverify.atlassian.net/browse/OJ-1922?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/unit/src/lib/validator.test.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  "AW283902A",
+  "AX283902A",
+  "AY283902A",
+  "AZ283902A",
+  "AA283902A",
+  "AA283902B",
+  "AA283902C",
+  "AA283902D",
+];
+const bad_ninos = [
```
Comment (surajkumar) [inline tests/unit/src/lib/validator.test.js:48]: Maybe we can expand on this list of bad ninos e.g. ninos with invalid lengths, invalid characters etc.

Comment (surajkumar) [inline tests/unit/src/lib/validator.test.js:48]: Maybe we can expand on this list of bad ninos e.g. ninos with invalid lengths, invalid characters etc.

File: src/lib/validator.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const nino = (value) => {
```
Comment (surajkumar) [inline src/lib/validator.js:1]: A more meaningful name would be better here e.g. `isValidNino`

**Review by surajkumar — CHANGES_REQUESTED** [review]
Can you check your PR please? It looks like new files are added that already exist?

**Comment by surajkumar** [comment]
Commits are not signed either!

File: tests/unit/src/lib/validator.test.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
+const { isValidNino } = require("../../../../src/lib/validator");
+
+const good_ninos = [
+  "AA283902A",
```
Comment (surajkumar) [inline tests/unit/src/lib/validator.test.js:4]: Question for @merlinc : Since this is a public repo and if by chance these are real NiNo's, would this be a problem?

**Review by merlinc — CHANGES_REQUESTED** [review]
Can I get a screenshot of what the frontend looks like when the invalid nino is entered?

**Comment by liamStockwellGDS** [comment]
> Can I get a screenshot of what the frontend looks like when the invalid nino is entered?

Screenshot 2023-11-01 at 17 13 35

**Comment by liamStockwellGDS** [comment]
> > Can I get a screenshot of what the frontend looks like when the invalid nino is entered?
> 
I've Updated the error messages.
Screenshot 2023-11-03 at 16 08 48
Screenshot 2023-11-03 at 16 08 39

File: tests/unit/src/lib/validator.test.js (hunk 3 above)
Comment (merlinc) [reply tests/unit/src/lib/validator.test.js:4]: Interesting question. We're using this list as known valid, but really the value is making sure that known incorrect ninos are detected. 

Maybe we can drop this list down to only a few entries?

File: src/lib/validator.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+const isValidNino = (value) => {
+  return value.match(
+    "" ||
+      /^(?!BG|GB|KN|NK|NT|TN|ZZ)[A-CEGHJ-PR-TW-Z][A-CEGHJ-NPR-TW-Z][0-9]{6}[ABCD]$/i
```
Comment (merlinc) [inline src/lib/validator.js:4]: This matches the format on HMRC's website  - https://www.gov.uk/hmrc-internal-manuals/national-insurance-manual/nim39110

Given they're big strings of letters, what do we think it more readable?
- `A-CEGHJ-PR-TW-Z`
- `ABCEGHJQLMNOPRSTWXZ`

File: tests/unit/src/lib/validator.test.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  "AW283902A",
+  "AX283902A",
+  "AY283902A",
+  "AZ283902A",
+  "AA283902A",
+  "AA283902B",
+  "AA283902C",
+  "AA283902D",
+];
+const bad_ninos = [
```
Comment (merlinc) [inline tests/unit/src/lib/validator.test.js:48]: Given the regex groups above, a potentially useful refactoring would be to break this up into the groups above, so we're clear about which bits we're testing. Otherwise it's a block of data being validated with no context.

**Review by merlinc — CHANGES_REQUESTED** [review]

File: tests/unit/src/lib/validator.test.js (hunk 3 above)
Comment (surajkumar) [reply tests/unit/src/lib/validator.test.js:4]: Yeah, I think dropping the number to a more concise list might be better.

File: src/lib/validator.js (hunk 4 above)
Comment (liamStockwellGDS) [reply src/lib/validator.js]: I think the big string way is more readable. I've also made this change.

**Review by merlinc — CHANGES_REQUESTED** [review]
PR contents LGTM

As a final bit of housekeeping can you squash the commits and rebase over main?
